OpenSim/Actuators/Millard2012EquilibriumMuscle.cpp

```diff
@@ -49,132 +49,126 @@ void Millard2012EquilibriumMuscle::constructProperties()
+    // from the parameters of the force-velocity curve.
+    double conSlopeAtVmax   = fvCurve.getConcentricSlopeAtVmax();
+    double conSlopeNearVmax = fvCurve.getConcentricSlopeNearVmax();
+    double isometricSlope   = fvCurve.getIsometricSlope();
+    double eccSlopeAtVmax   = fvCurve.getEccentricSlopeAtVmax();
+    double eccSlopeNearVmax = fvCurve.getEccentricSlopeNearVmax();
+    double conCurviness     = fvCurve.getConcentricCurviness();
+    double eccCurviness     = fvCurve.getEccentricCurviness();
+    double eccForceMax      = fvCurve.getMaxEccentricVelocityForceMultiplier();
+
+    // A few parameters may need to be adjusted to avoid singularities (e.g., if
+    // an elastic tendon is used with no fiber damping).
+    if(!get_ignore_tendon_compliance() && !use_fiber_damping) {
+        // Compliant tendon with no damping.
+        set_minimum_activation(clamp(0.01, get_minimum_activation(), 1));
+
+        if (falCurve.getMinValue() < 0.1)
+            falCurve.setMinValue(0.1);
+        if (conSlopeAtVmax < 0.1 || eccSlopeAtVmax < 0.1)
+            fvCurve.setCurveShape(0.1, conSlopeNearVmax, isometricSlope,
+                                  0.1, eccSlopeNearVmax, eccForceMax);
+
+    } else { //singularity-free model
+        set_minimum_activation(clamp(0, get_minimum_activation(), 1));
+        falCurve.setMinValue(0.0);
+        fvCurve.setCurveShape(0.0, conSlopeNearVmax, isometricSlope,
+                              0.0, eccSlopeNearVmax, eccForceMax);
+    }
 
-    } catch(const std::exception &x) {
-        std::string msg = "Exception caught in " + getName()
-                          + "::buildMuscle()\n" + x.what();
-        throw OpenSim::Exception(msg);
+    if (conSlopeAtVmax < 0.1 || eccSlopeAtVmax < 0.1) {
+        conSlopeAtVmax = 0.1;
+        eccSlopeAtVmax = 0.1;
+    }
+    fvInvCurve = ForceVelocityInverseCurve(conSlopeAtVmax, conSlopeNearVmax,
+                                           isometricSlope, eccSlopeAtVmax,
+                                           eccSlopeNearVmax, eccForceMax,
+                                           conCurviness, eccCurviness);
+
+    // Ensure all muscle curves are up-to-date.
+    falCurve.ensureCurveUpToDate();
+    fvCurve.ensureCurveUpToDate();
+    fvInvCurve.ensureCurveUpToDate();
+    fpeCurve.ensureCurveUpToDate();
+    fseCurve.ensureCurveUpToDate();
+
+    // Set properties of pennation model subcomponent. Values of
+    // optimal_fiber_length, pennation_angle_at_optimal, and
+    // maximum_pennation_angle are checked by
+    // MuscleFixedWidthPennationModel::extendFinalizeFromProperties().
+    auto& penMdl =
+        updMemberSubcomponent<MuscleFixedWidthPennationModel>(penMdlIdx);
+    penMdl.set_optimal_fiber_length(getOptimalFiberLength());
+    penMdl.set_pennation_angle_at_optimal(getPennationAngleAtOptimalFiberLength());
+    penMdl.set_maximum_pennation_angle(get_maximum_pennation_angle());
+    penMdl.finalizeFromProperties();
```

Comment:
I moved `penMdl.finalizeFromProperties();` down to the line that required it to be called, added a comment explaining as much, and removed `actMdl.finalizeFromProperties();` which was indeed unnecessary since it's called automatically by the Component machinery.